File: src/cc-h5/src/core/seed.ts
namespace core {
/////////////////////////////////////////////////////////////////////////////

export interface Seed
{
    head: Seed.Head;
    size: Seed.Size;
    cube: Array<Seed.Cube>;
    dest: Array<Seed.Vec2>;
}

export namespace Seed
{
    export interface Head
    {
        title: string;
        [ key: string]: string;
    }

    export interface Size
    {
         width: number;
        height: number;
    }

    export interface Vec2 extends Array<number>
    {
        /** x */
        0: number;
        /** y */
        1: number;
    }

    export interface Cube
    {
        type: "W"|"R"|"G"|"B";
        body: Array<Vec2>;
        move: Cube.Move|undefined;
    }

    export namespace Cube
    {
        export interface Move
        {
            loop: boolean;
            path: Path;
        }

        export interface Path extends Array<["I"|"L"|"D"|"U"|"R", number]>
        {}
    }
}

/////////////////////////////////////////////////////////////////////////////

export namespace Seed.Cube
{
    export import Type   = core.Cube.Type;
    export import Action = core.Cube.Action;

    export function toAction(src: string): Action
    {
        switch (src) {
            default:
            case "I": case "i": return Action.Idle;
            case "L": case "l": return Action.Left;
            case "D": case "d": return Action.Down;
            case "U": case "u": return Action.Up;
            case "R": case "r": return Action.Right;
        }
    }

    export function toActions(src: Path): Array<[Action, number]>
    {
        return src.map(([a, i]) => [toAction(a), i] as [Action, number]);
    }

    export function toType(src: string): Type
    {
        switch (src) {
            default:
            case "W": case "w": return Type.White;
            case "G": case "g": return Type.Green;
            case "B": case "b": return Type.Blue;
            case "R": case "r": return Type.Red;
        }
    }
}

/////////////////////////////////////////////////////////////////////////////

export namespace Guard
{
    export function isSeed(it: any): it is Seed
    {
        const its = it as Seed;
        return isHead(its.head)
            && isSize(its.size)
            && isCube(its.cube)
            && isDest(its.dest)
            ;
    }

    function isHead(it: any): it is Seed.Head
    {
        const its = it as Seed.Head;
        return typeof its.title === 'string';
    }

    function isSize(it: any): it is Seed.Size
    {
        const its = it as Seed.Size;
        return typeof its. width === 'number'
            || typeof its.height === 'number'
            ;
    }

    function isCube(it: any): it is Array<Seed.Cube>
    {
        return Array.isArray(it);
    }

    function isDest(it: any): it is Array<Seed.Vec2>
    {
        return Array.isArray(it);
    }
}

/////////////////////////////////////////////////////////////////////////////
}

File: tools/cc-edt/Editor/Models/Seed.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Editor.Models
{
    using Utilities;

    public class Seed
    {
        public Head head;
        public Size size;
        public List<Cube> cube;
        public List<Vec2> dest;

        public class Head : Dictionary<string, string>
        {

        }

        public struct Size
        {
            public int width;
            public int height;
        }

        public class Cube
        {
            [JsonConverter(typeof(StringEnumConverter))]
            public Type type;

            public List<Vec2> body;

            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
            public Move move;

            [JsonConverter(typeof(StringEnumConverter))]
            public enum Type
            {
                W = Models.Cube.Type.White,
                G = Models.Cube.Type.Green,
                B = Models.Cube.Type.Blue,
                R = Models.Cube.Type.Red,
            }

            public class Move
            {
                public bool loop;
                public Path path;

                public class Path : List<Pair>
                {

                }

                [JsonConverter(typeof(ObjectToArrayConverter<Pair>))]
                public struct Pair
                {
                    [JsonProperty(Order = 1)]
                    [JsonConverter(typeof(StringEnumConverter))]
                    public Type type;

                    [JsonProperty(Order = 2)]
                    public int time;

                    [JsonConverter(typeof(StringEnumConverter))]
                    public enum Type
                    {
                        I = Models.Cube.Path.Doing.Idle,
                        L = Models.Cube.Path.Doing.Left,
                        D = Models.Cube.Path.Doing.Down,
                        U = Models.Cube.Path.Doing.Up,
                        R = Models.Cube.Path.Doing.Right,
                    }
                }
            }
        }

        [JsonConverter(typeof(ObjectToArrayConverter<Vec2>))]
        public class Vec2
        {
            [JsonProperty(Order = 1)] public int x;
            [JsonProperty(Order = 2)] public int y;
        }
    }

    // Note:
    //
    // [Json.Net - Introduction]
    // (https://www.newtonsoft.com/json/help/html/Introduction.htm)
    // [Deserialize json tuple to object using ordinal index]
    // (https://stackoverflow.com/a/47963611)
    // [Ignoring null fields in Json.net]
    // (https://stackoverflow.com/a/22039616)
}
